etna/libs/timesfm/timesfm.py

```diff
@@ -0,0 +1,325 @@
+      incurred by, or claims asserted against, such Contributor by reason
+      of your accepting any such warranty or additional liability.
+"""
+
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Note: Copied from timesfm repository (https://github.com/google-research/timesfm/blob/154248137ccce29b01f4c3a765e85c3d9e4d92ba/src/timesfm/timesfm_torch.py)
+# Add method to change horizon after initialization.
+# Minor logic change of loading model.
+
+"""TimesFM pytorch forecast API for inference."""
+
+import logging
+from os import path
+from typing import Any, Sequence, Optional, Tuple
+import os
+import numpy as np
+import torch
+from huggingface_hub import snapshot_download
+from etna.libs.timesfm import timesfm_base
+
+from etna.libs.timesfm import patched_decoder as ppd
+
+_TOL = 1e-6
+
+
+class TimesFmTorch(timesfm_base.TimesFmBase):
+  """TimesFM forecast API for inference."""
+
+  def __post_init__(self):
+    self._model_config = ppd.TimesFMConfig(
+        num_layers=self.num_layers,
+        num_heads=self.num_heads,
+        hidden_size=self.model_dims,
+        intermediate_size=self.model_dims,
+        patch_len=self.input_patch_len,
+        horizon_len=self.output_patch_len,
+        head_dim=self.model_dims // self.num_heads,
+        quantiles=self.quantiles,
+        use_positional_embedding=self.use_pos_emb,
+    )
+    self._model = None
+    self.num_cores = 1
+    self.global_batch_size = self.per_core_batch_size
+    self._device = torch.device("cuda:0" if (
+        torch.cuda.is_available() and self.backend == "gpu") else "cpu")
+    self._median_index = -1
+
+  def _set_horizon(self, horizon):  # changed: added to change horizon after initialization
+      self.horizon_len = horizon
+
+  def load_from_checkpoint(
+      self,
+      checkpoint: timesfm_base.TimesFmCheckpoint,
+  ) -> None:
+    """Loads a checkpoint and compiles the decoder."""
+    checkpoint_path = checkpoint.path
+    repo_id = checkpoint.huggingface_repo_id
+    if not os.path.exists(checkpoint_path):  # changed: make loading similar to chronos
+      checkpoint_path = path.join(snapshot_download(checkpoint_path, cache_dir=checkpoint.local_dir), "torch_model.ckpt")
+    self._model = ppd.PatchedTimeSeriesDecoder(self._model_config)
+    loaded_checkpoint = torch.load(checkpoint_path)  # changed: remove weights_only=True due to attribute absence in low torch versions
```

Comment:
Does it mean we should do smth with it after update to `torch>2`?

Reply:
It is considered as a best practice to use `weights_only=True` but it will not brake anything in our case.
https://pytorch.org/tutorials/beginner/basics/saveloadrun_tutorial.html
We can, but it is unnecessary

Reply:
It will probably give us warning on torch>2. But we should see it on tests in the future.